### PR TITLE
Move Renderable and RenderableGroup to std::vector

### PR DESCRIPTION
**Pull request by trisyoungs**
This PR moves the `Renderable` and `RenderableGroup` classes on to `std::vector`, with the former stored as a `std:vector` of `std::shared_ptr`s. This also lays groundwork for other changes to the rendering system, which requires some modification once the `ObjectStore` class is removed.

`TreeWidgetUpdater` has been politely hacked with additional functions so as to not break the GUI, but this should be considered a temporary fix.

It also:
- Adds missing plot of improper energy to `EnergyModule`.
- Removes `Renderable` and `RenderableGroup` from depending on `List` and `ListItem` (#257).

### Related PR Hierarchy

Listed in order of review / merge ordering (rebasing required at each stage):

- ~#575 Remove state I/O for tabs, views, and charts~
- `THIS` Move `Renderable` and `RenderableGroup` to `std:vector`

### CODE REVIEW

File: src/gui/render/renderablegroupmanager.cpp [hunk 1]
```diff
@@ -28,105 +28,83 @@ void RenderableGroupManager::clear()
...
-                lowestId = colourId;
-        }
+        if (stockColourUsageCount_[colourId] < stockColourUsageCount_[lowestId])
+            lowestId = colourId;
+    }
 
-        renderableGroup = new RenderableGroup(name, (StockColours::StockColour)lowestId);
-        groups_.own(renderableGroup);
-        ++stockColourUsageCount_[lowestId];
+    auto &newGroup = groups_.emplace_back(name, (StockColours::StockColour)lowestId);
```
Comment (StephenSmith25) [inline src/gui/render/renderablegroupmanager.cpp:47]: A static cast vs a C style cast may be better here, i.e.`static_cast<StockColours::StockColour>(lowestId)`

File: src/gui/render/view.cpp [hunk 2]
```diff
@@ -909,16 +909,21 @@ void View::autoFollowData()
     // Only update the axes if one of the renderables transformed data has changed, to prevent needless primitive
     // regeneration further down the line
     auto updateRequired = false;
-    ListIterator<Renderable> renderableIterator(renderables_);
-    while (Renderable *rend = renderableIterator.iterate())
+    for (const auto rend : renderables_)
```
Comment (StephenSmith25) [inline src/gui/render/view.cpp:912]:
```suggestion
    for (const auto &rend : renderables_)
```

**Review by StephenSmith25 — COMMENTED** [review]
Looks good, just some minor comments and a couple of suggestions about passing shared pointers about by value vs reference vs reference to the object.

File: src/gui/render/renderablegroupmanager.cpp (hunk 1 above)
Comment (trisyoungs) [reply src/gui/render/renderablegroupmanager.cpp:47]: :+1:

File: src/gui/render/renderablegroup.cpp [hunk 3]
```diff
@@ -33,63 +57,68 @@ std::string_view RenderableGroup::name() const { return name_; }
...
     renderable->setValuesTransformEquation("value");
 
-    renderables_.remove(renderable);
-    renderable->setGroup(nullptr);
+    renderable->unSetGroup();
 }
 
 // Return whether the group is used by the specified renderable
-bool RenderableGroup::usedByRenderable(Renderable *renderable) const { return renderables_.contains(renderable); }
+bool RenderableGroup::usedByRenderable(std::shared_ptr<Renderable> renderable) const
```
[the remark this answers is not in the document]
Comment (trisyoungs) [reply src/gui/render/renderablegroup.cpp]: Yep. Have changed a few things around.

File: src/gui/render/renderablegroup.cpp [hunk 4]
```diff
@@ -33,63 +57,68 @@ std::string_view RenderableGroup::name() const { return name_; }
...
     setRenderableLineStyle(renderable);
 }
 
 // Remove Renderable from group (if it exists)
-void RenderableGroup::removeRenderable(Renderable *renderable)
+void RenderableGroup::removeRenderable(const std::shared_ptr<Renderable> renderable)
 {
-    if (!renderables_.contains(renderable))
+    // Return immediately if tne renderable is already in the group
+    if (std::find(renderables_.begin(), renderables_.end(), renderable) == renderables_.end())
```
[the remark this answers is not in the document]
Comment (trisyoungs) [reply src/gui/render/renderablegroup.cpp]: Rewritten for even more brevity!

**Review by StephenSmith25 — APPROVED** [review]
👍

**Review by StephenSmith25 — APPROVED** [review]
👍

**Pull request merged**